metarhia/metatests
# Refactor whole tests to remove cyclic deps and add functionality

* get rid of cyclic dependencies between modules

Namespaces:
* Removed and replaced with per-case namespace

Report:
* Refactor Reporter class with general reporting functionality
* Add ConciseReporter that just prints minimal needed info and is used
  by default
* 'report' method no longer triggers any tests - it only reports overall
  result
* 'addTest' method allows to add more tests to be reported by this
  reporter
* 'record' is called on each Test to parse/save the results of
  individual tests

Case:
* Refactor without globals and implicit dependencies on other modules
* Clean up code, avoid unnecessary checks
* Add DeclarativeTest that implements Test for declarative tests and
  uses runNow method to run what previously 'case' method did
* Add 'case' function that just creates DeclarativeTest and uses default
  reporter if not provided

Test:
* Refactor with es6 classes, clean up code and aliases
* Now each test is separate and therefore any test-file can be run as a
  separate program via just node (test will be by default run on
  nextTick if not disabled val 'run = false' option)
* Now all results are recorded including successful (this will allow to
  provide rich error reporting later on and avoid unnecessary variables
  to keep track of)
* 'notOk' is now an alias of 'assetNot' (previously it was what 'fail'
  does now)
* Add 'fail' to fail the test immediately with specified error message
* Add 'test' that adds a subtest to this test. This test will end
  until all subtests have finished.
  (has 2 aliases testSync and testAsync)
* 'test' method also has 'parallel' boolean option (in options) that
  indicates if it should be run in parallel, false by default
* Add 'beforeEach' method (test, callback) - it will be run before each
  subtest and its result will be passed to the subtest
  (test.test('caption', (subtest, arg1, arg2) => {}))
* Add 'afterEach' method (test, callback) - it will be run after each
  subtest and next (sequential) subtest will not be run until this
  method calls callback.
* Allow to listen on test finish event .on('done', (test) => {})
* Add 'error' event for errors that happed after test has ended
  (signature is (test, failMessage)). Example errors: check after end,
  end after end.
* Add test timeout (30s default) that will fail the test if it hasn't
  finished within the timeout time
* Test plan now properly checks and finished the test
* Tests can now have metadata
* Tests now store results of all checks in 'results'

Also, added a separate commit that shouldn't be landed if possible, it adds a plug to run current `case` tests in `common` etc that use global namespace without significant changes
